File: Code/ProjectGenerator/sources/projectGenerator.cpp
#include <iostream>
#include <fstream>
#include <String.h>
#include <unistd.h>

#include <sys/types.h>
#include <sys/stat.h>
#include <ctime>

#define WELCOME_TEXT "-------------------------------------------------\r\n\tS32 Design Studio Project Generator\r\n\tNXP internal use only                     \r\n-------------------------------------------------"
#define SEPARATOR    "-------------------------------------------------\r\n"

using namespace std;

/* Function Declarations */
void CreateFolder(const string folderName, const string projectPath);
void CreateProjectTree(const string projectName, const string projectPath);
void GenerateFile(const string projectName, const string projectPath, const string destinationPath, const string sourceFileName, const string destFileName );
void ReplaceStringInPlace(std::string& subject, const std::string& search, const std::string& replace);
void CreateProject(const string projectName, const string projectPath);
bool DirExists(string path);

/* Global variable where the current working directory is stored */
string g_cwd;

/*******************************************************************************
* Main function: Check if the parameters are correct and call the function     *
* that generates the project.                                                  *
*******************************************************************************/
int main(int argc, char ** argv)
{	
	char * tempPath = new char[255];
	getcwd(tempPath, 255);
	g_cwd.assign(tempPath);
	delete tempPath;

	if(argc > 1)
	{
		if(strcmp(argv[1], "-h") == 0)
		{
			cout<<"Usage: projectGenerator.exe project_name path_where_to_save"<<endl;	
		}
		else if(argc != 3)
		{
			cout<<"Invalid arguments. Use -h to get help"<<endl;
		} 
		else
		{
            cout<<WELCOME_TEXT<<endl;
            cout<<"\t- Project Name: "<<argv[1]<<endl<<"\t- Project Location: "<<argv[2]<<endl;
            cout<<SEPARATOR;
			CreateProject(argv[1], argv[2]);
			cout<<"Project Generation succedeed!"<<endl;
		}
	}
	else
	{
		cout<<"Invalid arguments. Use -h to get help"<<endl;
	}

	return 0;
}

void CreateProject(const string projectName, string projectPath)
{
	const string sourceDebugCnfNames[4] =  {"template Debug_FLASH JLink.launch", \
			   							    "template Debug_RAM JLink.launch", \
									        "template Debug_FLASH PEmicro.launch",\
									        "template Debug_RAM PEmicro.launch"};
									  
	string destDebugCnfNames[4] = sourceDebugCnfNames;
    
    const string sourceDoxyDocName = "template.dox";
    string destDoxyDocName = sourceDoxyDocName;

	if(!DirExists(projectPath))
	{
		cerr<<"Destination directory cannot be opened. Please check permissions or the path!"<<endl;
		exit(-2);
	}
	
	if((projectPath.c_str()[projectPath.size() - 1] != '\\'))
	{
		projectPath.append("\\");
	}
	
	CreateProjectTree(projectName, projectPath);
	GenerateFile(projectName, projectPath, "", ".project", ".project");
	GenerateFile(projectName, projectPath, "", ".cproject", ".cproject");
	GenerateFile(projectName, projectPath, "", "ProcessorExpert.pe", "ProcessorExpert.pe");
	GenerateFile(projectName, projectPath, "Sources", "main.c", "main.c");
	GenerateFile(projectName, projectPath, "", "Readme.htm", "Readme.htm");

    ReplaceStringInPlace(destDoxyDocName, "template", projectName);
    GenerateFile(projectName, projectPath, "Doxygen", sourceDoxyDocName, destDoxyDocName);
    
	for(int i = 0; i < 4; i++)
	{
		ReplaceStringInPlace(destDebugCnfNames[i], "template", projectName);
		GenerateFile(projectName, projectPath, "Debug_Configurations", sourceDebugCnfNames[i], destDebugCnfNames[i]);
	}
}

void CreateFolder(const string folderName, const string projectPath)
{
	string tempPath;
	tempPath = "mkdir";
	tempPath.append(" ");
	tempPath.append(projectPath);
	tempPath.append(folderName);

	system(tempPath.c_str());
}

void CreateProjectTree(const string projectName, const string projectPath)
{
	CreateFolder(projectName, projectPath);
	CreateFolder("Sources", projectPath+projectName+"\\");
	CreateFolder("Doxygen", projectPath+projectName+"\\");
	CreateFolder("Debug_Configurations", projectPath+projectName+"\\");

}

void GenerateFile(const string projectName, const string projectPath, const string destinationPath, const string sourceFileName, const string destFileName )
{
	ifstream fileBuf;
	ofstream outFile;
	string fileText;
	string filePath;
	string line;
    string prjNameBeautified = projectName;
    string * strTime;
    string * strDate;
	time_t rawTime;
    struct tm * timeInfo;
    char fmtTime[100];
    
    time(&rawTime);
    timeInfo = localtime(&rawTime);
    strftime(fmtTime, 100, "%Y-%m-%d, %H:%M", timeInfo);
    strTime = new string(fmtTime);
    
    strftime(fmtTime, 100, "%Y-%m-%d", timeInfo);
    strDate = new string(fmtTime);
    
    ReplaceStringInPlace(prjNameBeautified, "_", " ");
    
	try
		{
			filePath = g_cwd + "\\" + "template\\" + destinationPath + "\\" + sourceFileName;
			fileBuf.open(filePath.c_str());
			if(!fileBuf.is_open())
			{
				cerr<<"An error has occured during generation. Please check the paths and template directory!"<<endl;
				exit(-1);
			}
			while(!fileBuf.eof())
				{
					getline(fileBuf, line);
					fileText.append(line + "\n");
				}
			fileBuf.close();
			ReplaceStringInPlace(fileText, "${ProjectName}", projectName);
			ReplaceStringInPlace(fileText, "${ProjectName_Beautified}", prjNameBeautified);
			ReplaceStringInPlace(fileText, "${DateTime}", *strTime);
			ReplaceStringInPlace(fileText, "${Date}", *strDate);
			filePath = (projectPath + projectName + "\\" + destinationPath + "\\" + destFileName);
			outFile.open(filePath.c_str());
			if(!outFile.is_open())
			{
				cerr<<"An error has occured during generation. Please check the paths and template directory!"<<endl;
				exit(-1);
			}
			outFile.write(fileText.c_str(), fileText.size());
			outFile.close();
			delete strTime;
			delete strDate;
		}
	catch(int e)
		{
			cerr<<"An error has occured during generation. Please check the paths and template directory!"<<endl<<"Error code:"<<e<<endl;
		}
}

void ReplaceStringInPlace(std::string& subject, const std::string& search,
                          const std::string& replace)
{
	size_t pos = 0;
	while ((pos = subject.find(search, pos)) != std::string::npos)
		{
			subject.replace(pos, search.length(), replace);
			pos += replace.length();
		}
}

bool DirExists(const string path)
{
    struct stat info;

    if(stat( path.c_str(), &info ) != 0 )
        return false;
    else if(info.st_mode & S_IFDIR)
        return true;
    else
        return false;
}

File: Code/ProjectGenerator/sources/projectGenerator.py
import sys, os

################################################################################
# Main function: Check if the parameters are correct and call the functions    #
# that create the project tree.                                                #
################################################################################
def main(argv):
    try:
        if argv[0] == '-h':
            error('Usage: projectGenerator.py project_name path_where_to_save')
        else:
            try:
                projectName = argv[0]
                projectPath = argv[1]
                if not os.path.exists(projectPath):
                    error('Path is not correct!')
                else:
                    print('-------------------------------------------------\r\n\tS32 Design Studio Project Generator\r\n\tNXP internal use only                     \r\n-------------------------------------------------')
                    print('\t- Project name: {0}\r\n\t- Path to project: {1}\r\n'.format(projectName, projectPath))
                    createProjectTree(projectName, projectPath)
                    createFiles(projectName, projectPath)
                    print('----Project generation completed!----')
            except:
                pass
    except:
        error('Usage: projectGenerator.py project_name path_where_to_save')

################################################################################
# Error function: Print an error message and exit                              #
################################################################################            
def error(errorDescription):
    print(errorDescription)
    exit(0)

################################################################################
# createProjectTree function: Create the necessary project folders             #
################################################################################
def createProjectTree(projectName, projectPath):
    try:
        os.mkdir('{0}\\{1}'.format(projectPath, projectName))
        os.mkdir('{0}\\{1}\\Debug_Configurations'.format(projectPath, projectName))
        os.mkdir('{0}\\{1}\\Sources'.format(projectPath, projectName))
        os.mkdir('{0}\\{1}\\Doxygen'.format(projectPath, projectName))
    except:
        error('Cannot create folders\r\nProject generation failed!')

################################################################################
# createFiles function: Create the files                                       #
################################################################################
def createFiles(projectName, projectPath):
    try:
        # Open the .project file template
        projFile = open(os.getcwd() + '\\template\\' + '.project', 'r')
        projFileStr = projFile.read()
        projFile.close()
        
        # Replace project name and copy the new file into the new project
        newProjFile = open('{0}\\{1}\\.project'.format(projectPath, projectName), 'w')
        newProjFile.write(projFileStr.replace('${ProjectName}', projectName))
        newProjFile.close()
        
        # Open the .cproject file template
        cprojFile = open(os.getcwd() + '\\template\\' + '.cproject', 'r')
        cprojFileStr = cprojFile.read()
        cprojFile.close()
        
        # Replace project name and copy the new file into the new project
        newcProjFile = open('{0}\\{1}\\.cproject'.format(projectPath, projectName), 'w')
        newcProjFile.write(cprojFileStr.replace('${ProjectName}', projectName))
        newcProjFile.close()
        
        # Open the main.c file template
        sourceFile = open(os.getcwd() + '\\template\\Sources\\' + 'main.c', 'r')
        sourceFileStr = sourceFile.read()
        sourceFile.close()
        
        # Replace project name and copy the new file into the new project
        newsourceFile = open('{0}\\{1}\\Sources\\main.c'.format(projectPath, projectName), 'w')
        newsourceFile.write(sourceFileStr.replace('${ProjectName}', projectName))
        newsourceFile.close()

        # Define debug configurations name
        debugConfNames = ['template Debug_FLASH JLink.launch', 'template Debug_RAM JLink.launch', 'template Debug_FLASH PEmicro.launch', 'template Debug_RAM PEmicro.launch']
        
        # Replace project names and copy files
        for i in range(0, (len(debugConfNames))):
            debugFile = open(os.getcwd() + '\\template\\Debug_Configurations\\' + debugConfNames[i], 'r')
            debugFileStr = debugFile.read()
            debugFile.close()
            
            newdebugFile = open('{0}\\{1}\\Debug_Configurations\\'.format(projectPath, projectName) + debugConfNames[i].replace('template', projectName) , 'w')
            newdebugFile.write(debugFileStr.replace('${ProjectName}', projectName))
            newdebugFile.close()
            
        # Open the ProcessorExpert.pe file template
        pexFile = open(os.getcwd() + '\\template\\' + 'ProcessorExpert.pe', 'r')
        pexFileStr = pexFile.read()
        pexFile.close()
        
        # Replace project name and copy the new file into the new project
        newpexFile = open('{0}\\{1}\\ProcessorExpert.pe'.format(projectPath, projectName), 'w')
        newpexFile.write(pexFileStr.replace('${ProjectName}', projectName))
        newpexFile.close()   

        # Open the doxygen documentation file template
        doxFile = open(os.getcwd() + '\\template\\Doxygen\\' + 'template.dox', 'r')
        doxFileStr = doxFile.read()
        doxFile.close()
        
        # Replace project name and copy the new file into the new project
        newDoxFile = open('{0}\\{1}\\Doxygen\\{2}.dox'.format(projectPath, projectName, projectName), 'w')
        newDoxFile.write(doxFileStr.replace('${ProjectName}', projectName))
        newDoxFile.close()            
    except IOError as e:
        error(e)
if __name__ == "__main__":
    main(sys.argv[1:])
    
    
 